numba/np/polynomial/polynomial_core.py

```diff
@@ -0,0 +1,218 @@
+        with cgutils.early_exit_if(c.builder, stack, domain_native.is_error):
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+
+        window_obj = c.pyapi.object_getattr_string(obj, "window")
+        with cgutils.early_exit_if_null(c.builder, stack, window_obj):
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+        window_native = c.unbox(typ.window, window_obj)
+        c.pyapi.decref(window_obj)
+        with cgutils.early_exit_if(c.builder, stack, window_native.is_error):
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+
+        polynomial.coef = coef_native.value
+        polynomial.domain = domain_native.value
+        polynomial.window = window_native.value
+
+    return NativeValue(polynomial._getvalue(),
+                       is_error=c.builder.load(is_error_ptr))
+
+
+@box(types.PolynomialType)
+def box_polynomial(typ, val, c):
+    """
+    Convert a native polynomial structure to a Polynomial object.
+    """
+    ret_ptr = cgutils.alloca_once(c.builder, c.pyapi.pyobj)
+    fail_obj = c.pyapi.get_null_object()
+
+    with ExitStack() as stack:
+        polynomial = cgutils.create_struct_proxy(typ)(c.context, c.builder,
+                                                      value=val)
+        coef_obj = c.box(types.Array(types.double, 1, 'C'), polynomial.coef)
+        with cgutils.early_exit_if_null(c.builder, stack, coef_obj):
+            c.builder.store(fail_obj, ret_ptr)
+
+        domain_obj = c.box(typ.domain, polynomial.domain)
+        with cgutils.early_exit_if_null(c.builder, stack, domain_obj):
+            c.builder.store(fail_obj, ret_ptr)
+
+        window_obj = c.box(typ.window, polynomial.window)
+        with cgutils.early_exit_if_null(c.builder, stack, window_obj):
+            c.builder.store(fail_obj, ret_ptr)
+
+        class_obj = c.pyapi.unserialize(c.pyapi.serialize_object(Polynomial))
+        with cgutils.early_exit_if_null(c.builder, stack, class_obj):
+            c.pyapi.decref(coef_obj)
+            c.pyapi.decref(domain_obj)
+            c.pyapi.decref(window_obj)
+            c.builder.store(fail_obj, ret_ptr)
+
+        domain_dtype = numpy_support.as_dtype(typ.domain.dtype)
+        domain_dtype_num = c.pyapi.long(domain_dtype.num)
+        intp_dtype = numpy_support.as_dtype(types.intp)
+        intp_dtype_num = c.pyapi.long(intp_dtype.num)
+
+        pred = c.builder.icmp_signed("==", domain_dtype_num, intp_dtype_num)
+        with c.builder.if_else(pred) as (then, otherwise):
+            with then:
+                res1 = c.pyapi.call_function_objargs(class_obj, (coef_obj,))
+                c.builder.store(res1, ret_ptr)
+            with otherwise:
+                res3 = c.pyapi.call_function_objargs(class_obj, (coef_obj,
+                                                                 domain_obj,
+                                                                 window_obj))
+                c.builder.store(res3, ret_ptr)
```

Comment:
Rather than doing this, store in the PolynomialType if the ctor is for the one or the three arguments version. And here, just check this field.

```suggestion
        if typ.n_args == 1:
                res1 = c.pyapi.call_function_objargs(class_obj, (coef_obj,))
                c.builder.store(res1, ret_ptr)
        else:
                res3 = c.pyapi.call_function_objargs(class_obj, (coef_obj,
                                                                 domain_obj,
                                                                 window_obj))
                c.builder.store(res3, ret_ptr)
```